crates/core/src/network/connection_manager.rs

```diff
@@ -229,33 +352,51 @@ impl ConnectionManager {
                 None,
             ))
             .map_err(|_| HandshakeError::PeerDropped)?;
-        let connection = Connection::new(peer_node_id, selected_protocol_version, sender);
-
-        let should_register = matches!(
-            hello.direction(),
-            ConnectionDirection::Unknown
-                | ConnectionDirection::Bidirectional
-                | ConnectionDirection::Reverse
+        let connection = Connection::new(
+            peer_node_id,
+            selected_protocol_version,
+            hello.swimlane(),
+            sender,
         );
 
         // Register the connection.
-        let task_id = self.start_connection_reactor(
+        let router = guard.router.clone();
+
+        let existing_connection = if should_register {
+            guard.register(connection.clone())
```

Comment:
Aren't we registering connection twice? Once here and once through the `ConnectionTracking`?

Reply:
Yes. Intentionally though. In this case, the other one can't put back the original connection. This only impacts incoming connections via accept() though.